### PR TITLE
Extract Go and Python Beam symbols for Playground

### LINKED ISSUE #23304: [Task]: Load Python Beam symbols dictionary into Playground
### What needs to happen?

Load Python Beam symbols dictionary into Playground

### Issue Priority

Priority: 3

### Issue Component

Component: beam-playground

### LINKED ISSUE #23534: [Task]: Version-control all generated Playground files
### What needs to happen?

1. Version-control all generated files.
2. Exclude all generated files from Apache Rat check (this really is the most painful thing with their version-controlling as we had been adding licenses manually there).
3. When building a Docker container, delete and re-generate all generated files so that their manual modifications have no effect in deployment and we see errors stemming from that early.
4. Add a Gradle task to delete all generated files so that generation is easy to verify and we can test the generated files' consistency before deployment.
5. Delete and re-generate all generated files in `:playground:frontend:precommit`
6. Add a GitHub post-commit workflow to re-generate files and to compare them with the committed versions. Break if a difference is found. This way the committer will receive an email and will soon make it a habit to run the local pre-commit task.

### Issue Priority

Priority: 3

### Issue Component

Component: beam-playground

### LINKED ISSUE #23917: [Task]: Extract Go Beam symbols for Playground
### What needs to happen?

Parse symbols so they can be suggested by the editor autocompletion.

See https://github.com/apache/beam/issues/23304 as a model.

### Issue Priority

Priority: 3

### Issue Component

Component: beam-playground

### PR DESCRIPTION
**Pull request by alexeyinkin**
This PR:
1. Adds a Gradle task to do all post-checkout tasks.
2. Adds a Python script to extract Beam symbols from Python SDK.
3. Adds a Gradle task to call that script and to place the symbols so that the frontend app can use them.
4. Updates a frontend Dockerfile to do the same.
5. Updates the README explaining those changes.

The symbols are not yet used by the app after loading. The app will start using them when the Akvelon editor replaces the old editor. Still this PR is needed for easier testing of the Akvelon editor before the switch and to simplify. It also simplifies post-checkout tasks.

- Resolves #23304
- Resolves #23917
- Resolves #23534

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

File: playground/frontend/playground_components/lib/src/playground_components.dart
Comment (Malarg) [inline playground/frontend/playground_components/lib/src/playground_components.dart:39]: Are we definitely have to add new DI system in our app? Why can't we use provider for the same goal?

**Review by Malarg — CHANGES_REQUESTED** [review]

File: playground/frontend/playground_components/lib/src/services/symbols/symbols_notifier.dart
Comment (alexeyinkin) [reply playground/frontend/playground_components/lib/src/services/symbols/symbols_notifier.dart:27]: A listener is in a WIP that will replace the editor with ours.

File: playground/frontend/Dockerfile
Comment (vchunikhin) [inline playground/frontend/Dockerfile:50]: Why do we use checkout here? It looks that it's not necessary

File: playground/frontend/playground_components/lib/src/services/symbols/loaders/yaml.dart
Comment (alexeyinkin) [reply playground/frontend/playground_components/lib/src/services/symbols/loaders/yaml.dart:62]: No, filed: https://github.com/apache/beam/issues/23398

File: playground/frontend/playground_components/lib/src/services/symbols/loaders/yaml.dart
Comment (alexeyinkin) [reply playground/frontend/playground_components/lib/src/services/symbols/loaders/yaml.dart:51]: Less conditions are easier to maintain. There could be a performance gain with `if` here, but this code is not executed that often.

Also such cases subject to compiler optimizations. If not now, then in the future.

File: playground/frontend/playground_components/lib/src/services/symbols/symbols_notifier.dart
Comment (alexeyinkin) [reply playground/frontend/playground_components/lib/src/services/symbols/symbols_notifier.dart:53]: It's in the WIP that will replace the editor with ours.

File: playground/frontend/playground_components/lib/src/playground_components.dart
Comment (alexeyinkin) [reply playground/frontend/playground_components/lib/src/playground_components.dart:39]: This is more of a service locator than a dependency injection. SL's objects are a form of global variables while provider is a mean to make something available to a subtree.

Provider should be used if it makes sense to override an object to certain parts of the tree. For instance, we do this with themes in 'Share my code' button.

For objects that are global by design, hanging them with provider adds complexity to their management. Also with SL objects are available in other business logic objects and not only in widgets.

`SymbolsNotifier` will be used in `SnippetEditingController`. Passing it though widgets is way more complex than just using a global instance with `get_it`.

File: playground/frontend/playground_components/lib/src/services/symbols/loaders/yaml.dart
Comment (alexeyinkin) [reply playground/frontend/playground_components/lib/src/services/symbols/loaders/yaml.dart]: Done.

File: playground/frontend/playground_components/lib/src/services/symbols/symbols_notifier.dart
Comment (alexeyinkin) [reply playground/frontend/playground_components/lib/src/services/symbols/symbols_notifier.dart]: Done.

File: playground/frontend/playground_components/lib/src/services/symbols/loaders/yaml.dart
Comment (alexeyinkin) [reply playground/frontend/playground_components/lib/src/services/symbols/loaders/yaml.dart:25]: Added tests.

File: playground/frontend/playground_components/test/src/services/symbols/symbols_notifier_test.dart
Comment (Malarg) [inline playground/frontend/playground_components/test/src/services/symbols/symbols_notifier_test.dart:55]: irrelevant comment

File: playground/frontend/playground_components/test/src/services/symbols/symbols_notifier_test.dart
Comment (Malarg) [inline playground/frontend/playground_components/test/src/services/symbols/symbols_notifier_test.dart:41]: Why is it necessary here? Provide comment, please

File: playground/frontend/playground_components/test/src/services/symbols/symbols_notifier_test.dart
Comment (Malarg) [inline playground/frontend/playground_components/test/src/services/symbols/symbols_notifier_test.dart:49]: this class should be private

File: playground/frontend/playground_components/test/src/services/symbols/loaders/yaml_test.dart
Comment (Malarg) [inline playground/frontend/playground_components/test/src/services/symbols/loaders/yaml_test.dart:37]: Why don't you follow rule "one check per one test"? It will be more clear to split this test to 3, and join them into group

**Review by Malarg — CHANGES_REQUESTED** [review]

File: playground/frontend/playground_components/test/src/services/symbols/symbols_notifier_test.dart
Comment (alexeyinkin) [reply playground/frontend/playground_components/test/src/services/symbols/symbols_notifier_test.dart:41]: Loaders are asynchronous. `TestLoader` has `future` that returns an object without awaiting, but it is still awaited in `SymbolsNotifier`. That `await` yields the control. To resume loading, the main test thread has to yield in turn.

I don't think a comment in the code is required here because `await Future.delayed(Duration.zero)` is a common pattern in Dart and is self-documenting.

File: playground/frontend/playground_components/test/src/services/symbols/loaders/yaml_test.dart
Comment (alexeyinkin) [reply playground/frontend/playground_components/test/src/services/symbols/loaders/yaml_test.dart:37]: That way the 'Act' in the AAA would be repeated across those tests. I don't like such code duplication. A way to avoid this would be to put both the Arrange and Act into `setUp()`, but that is more confusing because `setUp` is designed for Arrange only.

File: playground/frontend/playground_components/test/src/services/symbols/symbols_notifier_test.dart
Comment (alexeyinkin) [reply playground/frontend/playground_components/test/src/services/symbols/symbols_notifier_test.dart]: Done.

File: playground/frontend/playground_components/test/src/services/symbols/symbols_notifier_test.dart
Comment (alexeyinkin) [reply playground/frontend/playground_components/test/src/services/symbols/symbols_notifier_test.dart]: Deleted.

File: playground/frontend/playground_components/lib/src/services/symbols/loaders/yaml.dart
Comment (alexeyinkin) [reply playground/frontend/playground_components/lib/src/services/symbols/loaders/yaml.dart]: Done.

File: playground/frontend/playground_components/test/src/services/symbols/symbols_notifier_test.dart
Comment (Malarg) [reply playground/frontend/playground_components/test/src/services/symbols/symbols_notifier_test.dart:41]: Thanks

**Review by Malarg — APPROVED** [review]

**Comment by alexeyinkin** [comment]
R: @damondouglas

File: playground/frontend/README.md
Comment (damondouglas) [inline playground/frontend/README.md:55]: Is Docker the only way to run this application?  Is there any way that we can include instructions for developers who want to contribute, but who are prevented for various reasons from using Docker?

May we consider adding instructions for these developers, including where it is expected to execute `flutter pub run build_runner build` such as in the playground_components folder?  I'm assuming such command runs in the context of the gradle command to setup the project, however, please see the errors I encountered when executing this setup command.

File: playground/frontend/playground_components/tools/extract_symbols_python/extract_python_symbols.py
Comment (damondouglas) [inline playground/frontend/playground_components/tools/extract_symbols_python/extract_python_symbols.py:26]: Is there a plan for unit tests for this method as well as the other methods?

**Review by damondouglas — CHANGES_REQUESTED** [review]

File: playground/frontend/README.md
Comment (alexeyinkin) [reply playground/frontend/README.md:34]: With Cloud Shell, I narrowed down the problem and filed an issue:
https://issuetracker.google.com/issues/251039262

Can you please check if `git clone --mirror` also fails for you? If so, please upvote the issue and maybe provide additional details there (project ID?).

File: playground/frontend/README.md
Comment (alexeyinkin) [reply playground/frontend/README.md:65]: Docker is not required. The recommended way to run this locally is still `flutter run`.
I updated the README stating the cases when one might prefer Docker.

Comment (alexeyinkin) [reply playground/frontend/README.md:65]: As for the local machine, I cannot reproduce this, and I cannot really see how is Flutter in $PATH for the terminal but not in $PATH for Gradle ran from that same terminal. Can you show the output of these?

1. `cat /etc/*-release`
2. `which flutter`
3. `echo $PATH`
4. `./gradlew :playground:frontend:printPath` (a new task I just added)

File: playground/frontend/README.md
Comment (alexeyinkin) [reply playground/frontend/README.md]: As for the local machine, I cannot reproduce this, and I cannot really see how is Flutter in $PATH for the terminal but not in $PATH for Gradle ran from that same terminal. Can you show the output of these?

1. `cat /etc/*-release`
2. `which flutter`
3. `echo $PATH`
4. `./gradlew :playground:frontend:printPath` (a new task I just added)

File: playground/frontend/playground_components/tools/extract_symbols_python/extract_python_symbols.py
Comment (alexeyinkin) [reply playground/frontend/playground_components/tools/extract_symbols_python/extract_python_symbols.py]: Done. I made a test for the entire command in Dart.

First I thought about testing Python code in Python, but it has a drawback. We will have 3 more scripts for this job in other languages, very much different inside but with an identical API. In Dart we can make 4 nearly identical tests instead of 4 very different native tests on those languages. So the Dart solution is easier to maintain.

File: playground/frontend/README.md
Comment (damondouglas) [reply playground/frontend/README.md]: Running the above commands will show I tested on Linux Debian and that flutter is in my path.  I don't want to delay this PR further and thank you for this information.  However, running `./gradlew :playground:frontend:configure` resulted in a `Module not found yaml` error which I resolved by installing the missing python library.  Now `./gradlew :playground:frontend:configure --stacktrace` reveals this error:

```
> Task :playground:frontend:playground_components:extractBeamSymbolsPython FAILED
Traceback (most recent call last):
  File "playground/frontend/playground_components/tools/extract_symbols_python/extract_symbols_python.py", line 109, in <module>
    class_names = get_dir_symbols_recursive(args.dir)
  File "playground/frontend/playground_components/tools/extract_symbols_python/extract_symbols_python.py", line 100, in get_dir_symbols_recursive
    class_names.update(get_file_symbols(file_path))
  File "playground/frontend/playground_components/tools/extract_symbols_python/extract_symbols_python.py", line 66, in get_file_symbols
    [target] = member_node.targets
ValueError: too many values to unpack (expected 1)

```

File: playground/frontend/README.md
Comment (damondouglas) [inline playground/frontend/README.md:37]: Does this PR still intend on deleting https://github.com/apache/beam/blob/master/playground/frontend/README.md?plain=1#L32-L37?

**Review by damondouglas — COMMENTED** [review]
@pabloem I wasn't able to fully review this PR but didn't want to hold it back.

Pros:
1. flutter code is readable and `flutter run` results in a working frontend
2. `flutter test` passes

Opportunities:
1. Not clear whether https://github.com/apache/beam/blob/master/playground/frontend/README.md?plain=1#L32-L37 will remain after this PR will merge and personally I think it helps to have these instructions
2. `./gradlew :playground:frontend:configure` kept throwing errors and I can't rule out whether it is my execution environment

**Comment by alexeyinkin** [comment]
It's OK to hold this PR until everything builds on any machine. This does not block us because meanwhile we are busy on the editor.

We expect to migrate to Akvelon editor next week. This means we will no longer have a git dependency which means Google Cloud Shell will start working.

Meanwhile I will fix the Python script using your feedback and list `pyyaml` in the README.

File: playground/frontend/README.md
Comment (alexeyinkin) [reply playground/frontend/README.md:37]: I deleted these lines because with the introduction of symbols extraction the building got more complicated. That is why `:playground:frontend:configure` was introduced.

With the recent idea of keeping generated files under version control, `:playground:frontend:configure` may remain as simple as the above 6 commands. This means it can hardly break, so I guess it should be safe to delete them from the REAMDE. Otherwise we will have two locations to maintain the build commands:
1. `:playground:frontend:configure` Gradle task.
2. README.

Yet the failure of Gradle on your machine with Flutter in path is a concern.

Do you want to keep those commands as a backup?

**Comment by damccorm** [comment]
stop reviewer notifications

**Comment by pabloem** [comment]
can you rebase this PR please @alexeyinkin ?

**Comment by pabloem** [comment]
oh I realize this was not lgtmd? @damondouglas should I review this? merge? waqit for your review?

**Comment by alexeyinkin** [comment]
@damondouglas now that we use the editor from pub.dev instead of git, we do not hit the bug of Google Cloud Shell, and it works all the way from `./gradlew :frontend:playground:configure` to web preview.

File: playground/frontend/playground_components/pubspec.yaml
Comment (alexeyinkin) [inline playground/frontend/playground_components/pubspec.yaml:65]: `**/generated/*` is git-ignored on the repository level, so we evade that.

**Comment by alexeyinkin** [comment]
Meanwhile I also did
- #23534
in this PR so as not to be blocked by this.

**Review by damondouglas — COMMENTED** [review]
@alexeyinkin Would the following remain an issue if we merged this PR?

### 1. I noticed that the text for some of the buttons/dropdowns were misaligned:

7BihGGZZUqo8MuS

and

75zQRqyE98nVfNG

### 2. Possibly related to 1. There appears to be a size overflow error.
8JWfeMcPSe72gD8

**Comment by alexeyinkin** [comment]
@damondouglas how did you build and run it to get that? My tests:

## 1. Cloud Shell

```bash
cp playground/frontend/lib/config.example.dart playground/frontend/lib/config.g.dart
cd playground/frontend/
docker build -t playground-frontend .
docker run -p 8080:8080 playground-frontend
```
And then Web Preview:
image

Flutter 3.3.2, Chrome.

## 2. Local debug with Android Studio

```bash
cp playground/frontend/lib/config.example.dart playground/frontend/lib/config.g.dart
```

image

Flutter 3.3.7, Chrome.

**Comment by damondouglas** [comment]
Good thank you.  I executed the commands locally in the playground/frontend/README on a Linux Debian machine.  I'm not remoting into the device currently to inform the flutter version but updated prior to execution.  I'd say then this PR is good to go.

**Review by damondouglas — APPROVED** [review]
@pabloem LGTM

**Comment by pabloem** [comment]
lgtm thanks

**Pull request merged**
